File: FacadeDesignPattern/JavaScript/facade.js
// Subsystems
class PowerSupply {
  providePower() {
    console.log("Power Supply: Providing power...");
  }
}

class CoolingSystem {
  startFans() {
    console.log("Cooling System: Fans started...");
  }
}

class CPU {
  initialize() {
    console.log("CPU: Initialization started...");
  }
}

class Memory {
  selfTest() {
    console.log("Memory: Self-test passed...");
  }
}

class HardDrive {
  spinUp() {
    console.log("Hard Drive: Spinning up...");
  }
}

class BIOS {
  boot(cpu, memory) {
    console.log("BIOS: Booting CPU and Memory checks...");
    cpu.initialize();
    memory.selfTest();
  }
}

class OperatingSystem {
  load() {
    console.log("Operating System: Loading into memory...");
  }
}

// Facade
class ComputerFacade {
  constructor() {
    this.power = new PowerSupply();
    this.cool = new CoolingSystem();
    this.cpu = new CPU();
    this.memory = new Memory();
    this.hard = new HardDrive();
    this.bios = new BIOS();
    this.ops = new OperatingSystem();
  }

  startComputer() {
    console.log("<----------------- Computer starting ------------->");
    this.power.providePower();
    this.cool.startFans();
    this.cpu.initialize();
    this.memory.selfTest();
    this.hard.spinUp();
    this.bios.boot(this.cpu, this.memory);
    this.ops.load();
    console.log("<--------------- Computer started -------------->");
  }
}

// Client
const computer = new ComputerFacade();
computer.startComputer();

File: FacadeDesignPattern/Cpp/facade.cpp
#include<bits/stdc++.h>
using namespace std;


// Subsystems
class PowerSupply {
public:
    void providePower() {
        cout << "Power Supply: Providing power..." << endl;
    }
};
class CoolingSystem {
public:
    void startFans() {
        cout << "Cooling System: Fans started..." << endl;
    }
};
class CPU {
public:
    void initialize() {
        cout << "CPU: Initialization started..." << endl;
    }
};


class Memory {
public:
    void selfTest() {
        cout << "Memory: Self-test passed..." << endl;
    }
};

class HardDrive {
public:
    void spinUp() {
        cout << "Hard Drive: Spinning up..." << endl;
    }
};

class BIOS {
public:
    void boot(CPU& cpu, Memory& memory) {
        cout << "BIOS: Booting CPU and Memory checks..." << endl;
        cpu.initialize();
        memory.selfTest();
    }
};

class OperatingSystem {
public:
    void load() {
        cout << "Operating System: Loading into memory..." << endl;
    }
};

//Facade

class ComputerFacade{
private: 
    PowerSupply power;
     CoolingSystem cool;
     CPU cpu;
     Memory memory;
     HardDrive hard;
     BIOS bios;
     OperatingSystem ops;
public:    void startComputer(){
        cout<<"<----------------- Computer starting ------------->"<<endl;
        power.providePower();
        cool.startFans();
        cpu.initialize();
        memory.selfTest();
        hard.spinUp();
        bios.boot(cpu,memory);
        ops.load();
        cout<<"<--------------- Computer started -------------->"<<endl;
    }

};

//Client
int main(){
    ComputerFacade computer;
    computer.startComputer();
    return 0;

}